lib/rules/indent.js

```diff
@@ -947,6 +945,36 @@ module.exports = {
             return !node || node.range[0] === token.range[0];
         }
 
+        /**
+         * Check whether the given tokens are pairs of "{" and "}"
+         * @param {Token} openBraceToken is "[".
+         * @param {Token} closeBraceToken is "]".
+         * @returns {boolean} `true` if the two tokens are complementary pair
+         */
+        function isComplementBrace(openBraceToken, closeBraceToken) {
+            return astUtils.isOpeningBraceToken(openBraceToken) && astUtils.isClosingBraceToken(closeBraceToken);
+        }
+
+        /**
+         * Check whether the given tokens are pairs of "]" and "[", in reverse
+         * @param {Token} openBracketToken is "{".
+         * @param {Token} closeBracketToken is "}".
+         * @returns {boolean} `true` if the two tokens are complementary pair
+         */
+        function isComplementBracket(openBracketToken, closeBracketToken) {
+            return astUtils.isOpeningBracketToken(openBracketToken) && astUtils.isClosingBracketToken(closeBracketToken);
+        }
+
+        /**
+         * Check whether the given tokens are pairs of "{" and "}" or "]" and "["
+         * @param {Token} openToken is "{" or "[".
+         * @param {Token} closeToken is "}" or "]".
+         * @returns {boolean} `true` if the two tokens are complementary pair
+         */
+        function isComplementBraceOrBracket(openToken, closeToken) {
+            return isComplementBrace(openToken, closeToken) || isComplementBracket(openToken, closeToken);
+        }
```

Comment:
It seems like your logic relies a lot on finding matching brackets for destructuring. However, I'm not sure this is a good approach -- I don't think the problem is specific to whether the brackets match. For example, the rule will also incorrectly report an error for:

```js
const [
    foo
] = bar(
    baz
);
```